lxd/device/device.go

```diff
@@ -0,0 +1,115 @@
+	// VolatileConfig from previous instantiations of this device is required.
+	// It calls the supplied volatile save function if state needs persisting.
+	Stop(VolatileConfig, func(map[string]string) error) error
+
+	// Update performs live host-side modifications for a started device based on the difference
+	// between the current config and previous config supplied as an argument.
+	// VolatileConfig from previous instantiations of this device is required.
+	// It calls the supplied volatile save function if state needs persisting.
+	Update(VolatileConfig, Config, func(map[string]string) error) error
+
+	// UpdateFields returns a list of fields that can be updated whilst the device is started.
+	UpdateFields() []string
+
+	// CanHotPlug returns true if device can be started/stopped whilst instance is running.
+	CanHotPlug() bool
+}
+
+// device represents a sealed interface that implements Device, but also contains some internal
+// setup functions for a Device that should only be called by device.New() to avoid exposing devices
+// that are not in a known configured state. This is separate from the Device interface so that
+// Devices created outside of the device package can be used by LXD, but ensures that any devices
+// created by the device package will only be accessible after being configured properly by New().
+type device interface {
+	Device
+
+	// init stores the instanceIdentifier, daemon State and Config into device and performs any setup.
+	init(instanceIdentifier, *state.State, Config)
+
+	// validate checks Config stored by init() is valid for the instance type.
+	validate() error
+}
+
+// deviceCommon represents the common struct for all devices.
+type deviceCommon struct {
+	instance instanceIdentifier
+	config   map[string]string
+	state    *state.State
+}
+
+// init stores the instanceIdentifier, daemon state and Config into device. This is implemented as
+// part of deviceCommon so that the majority of devices don't need to implement it and can just
+// embed deviceCommon.
+func (d *deviceCommon) init(instance instanceIdentifier, state *state.State, conf Config) {
+	d.instance = instance
+	d.config = conf
+	d.state = state
+}
+
+// Update returns an error as most devices do not support live updates.
+func (d *deviceCommon) Update(v VolatileConfig, oldConfig Config, save func(map[string]string) error) error {
+	return fmt.Errorf("Device does not support updates whilst started")
+}
+
+// UpdateFields returns an empty list of update fields as most devices do not support live updates.
+func (d *deviceCommon) UpdateFields() []string {
+	return []string{}
+}
+
+// CanHotPlug returns true as majority of devices can be started/stopped when instance is running.
+func (d *deviceCommon) CanHotPlug() bool {
+	return true
+}
+
+// New instantiates a new device struct, validates the supplied config and sets it into the device.
+func New(instance instanceIdentifier, state *state.State, devConfig Config) (Device, error) {
```

Comment:
Wouldn't it make more sense to pass the volatile get/set functions here, record them in the struct and have the individual functions check if they're set when needed and throw an error if they're not?

This would effectively eliminate the need for most arguments for the functions on the interface.

Reply:
@stgraber yep makes sense, I've added that now.